### PR TITLE
Implement SignedOperationProxy

### LINKED ISSUE #273: Signed Authorization Proxy [7]
Build a proxy that will allow actions to be performed if a user signs them

### PR DESCRIPTION
**Pull request by BrendanChou**
Fixes #273

### CODE REVIEW

File: src/modules/operate/AccountOperation.ts [hunk 1]
```diff
@@ -381,6 +392,117 @@ export class AccountOperation {
...
+    },
+  ): Operation {
+    if (this.proofs.length) {
+      throw new Error('Cannot create operation out of operation with proofs');
+    }
+    if (!this.actions.length) {
+      throw new Error('Cannot create operation out of operation with no actions');
+    }
+
+    function toEnum(input: string | number) {
```
Comment (AntonioJuliano) [inline src/modules/operate/AccountOperation.ts]: nit: don't declare functions in functions

File: src/modules/operate/AccountOperation.ts [hunk 2]
```diff
@@ -381,6 +392,117 @@ export class AccountOperation {
...
+            value: action.amount.value.times(action.amount.sign ? 1 : -1),
+          },
+        },
+      );
+    }
+
+    return this;
+  }
+
+  public createOperation(
```
Comment (AntonioJuliano) [inline src/modules/operate/AccountOperation.ts]: I don't understand what this function is for, can you explain?

File: src/modules/operate/AccountOperation.ts [hunk 3]
```diff
@@ -381,6 +392,117 @@ export class AccountOperation {
...
+            denomination: action.amount.denomination,
+            value: action.amount.value.times(action.amount.sign ? 1 : -1),
+          },
+        },
+      );
+    }
+
+    return this;
+  }
+
```
Comment (AntonioJuliano) [inline src/modules/operate/AccountOperation.ts]: I was imagining there would be `public sign` function which would take the current operation, sign it (create the proof), and return a `SignedOperation`

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+        ")"
+    );
+
+    // TODO: comment
+    bytes constant internal EIP712_OPERATION_STRING = abi.encodePacked(
+        "Operation(",
+        "Action[] actions,",
+        "uint256 expiration,",
+        "uint256 salt,",
+        "address sender",
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: yep, it's implemented!

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+    // Hash of the EIP712 Domain Separator data
+    bytes32 public EIP712_DOMAIN_HASH;
+
+    // ============ Mutable Storage ============
+
+     // true if this contract can process operationss
+    bool public g_isOperational;
+
+    // signer => final hash => was executed (or cancelled)
+    mapping (address => mapping (bytes32 => bool)) public g_invalidated;
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: it sets the variable to true if it's already executed as well

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+
+    // Hash of the EIP712 Domain Separator data
+    bytes32 public EIP712_DOMAIN_HASH;
+
+    // ============ Mutable Storage ============
+
+     // true if this contract can process operationss
+    bool public g_isOperational;
+
+    // signer => final hash => was executed (or cancelled)
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: true 🇺🇸

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+        // cache the index of the first action for this proof
+        uint256 proofStartIndex = 0;
+
+        // loop over all proofs
+        for (uint256 p = 0; p < proofs.length; p++) {
+            Proof memory proof = proofs[p];
+
+            // require that the message is not expired
+            Require.that(
+                proof.expiration == 0 || proof.expiration >= block.timestamp,
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: We did the same for limit orders yeah

File: src/modules/operate/AccountOperation.ts (hunk 2 above)
Comment (BrendanChou) [reply src/modules/operate/AccountOperation.ts]: it was for your comment below

File: src/modules/operate/AccountOperation.ts (hunk 3 above)
Comment (BrendanChou) [reply src/modules/operate/AccountOperation.ts]: There is the function `createOperation()` which creates the operation (unsigned) which can then be signed using any of the sign methods in `solo.signedOperations`. If you combined the sign message into this one then you would have to create several different functions in `AccountOperation` to use each kind of signing method

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+            g_isOperational,
+            FILE,
+            "Contract is not operational"
+        );
+
+        // cache the index of the first action for this proof
+        uint256 proofStartIndex = 0;
+
+        // loop over all proofs
+        for (uint256 p = 0; p < proofs.length; p++) {
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: Yes you will have to submit an empty proof. Which makes it slightly annoying on the caller of the function, but makes it significantly easier (and less bug prone in my opinion) to verify this stuff in solidity

File: docs/usage.md [hunk 9]
```diff
@@ -29,7 +29,7 @@ const operation = solo.operation.initiate();
 Solo also provides a [Payable Proxy](https://github.com/dydxprotocol/solo/blob/master/contracts/external/proxies/PayableProxyForSoloMargin.sol) contract that will automatically wrap and unwrap ETH <-> WETH, so that users can interact with Solo using only ETH. You can use it by:
 
 ```javascript
-const operation = solo.operation.initiate({ usePayableProxy: true });
+const operation = solo.operation.initiate({ proxy: ProxyType.Payable });
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply docs/usage.md]: Put in some code to allow existing js clients to not break, and tested it in the PayableProxy tests. ts clients may still break on compilation and they can either cast it to the correct type or update the code

File: src/lib/BytesHelper.ts [hunk 10]
```diff
@@ -34,12 +39,15 @@ export function addressToBytes32(input: address) {
...
-export function stringToBytes32(input: string) {
+export function hashString(input: string) {
   return soliditySha3({ t: 'string', v: input });
 }
 
-export function bytesToBytes32(input: string) {
-  return soliditySha3({ t: 'bytes', v: input });
+export function hashBytes(input: string) {
+  if (!stripHexPrefix(input)) {
+    return '0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470';
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply src/lib/BytesHelper.ts]: added comment, but basically soliditySha3 the javascript function is not great

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+        address signer
+    )
+        private
+        view
+    {
+        bool valid =
+            msg.sender == accountOwner
+            || signer == accountOwner
+            || SOLO_MARGIN.getIsLocalOperator(accountOwner, msg.sender)
+            || SOLO_MARGIN.getIsLocalOperator(accountOwner, signer);
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: Global operators already have full control so they would not need to use this I don't think

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+        emit LogOperationCanceled(operationHash, msg.sender);
+    }
+
+    /**
+     * TODO: description
+     */
+    function operate(
+        Account.Info[] memory accounts,
+        Actions.ActionArgs[] memory actions,
+        Proof[] memory proofs
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: Sure why not

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+    )
+        public
+        nonReentrant
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: Good point

[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: Removed

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 14]
```diff
@@ -0,0 +1,550 @@
...
+
+            // require that the sender is the sender
+            Require.that(
+                auth.sender == address(0) || auth.sender == msg.sender,
+                FILE,
+                "Operation sender mismatch",
+                p
+            );
+
+            // get the signer of the auth (msg.sender if no auth)
```
Comment (AntonioJuliano) [inline contracts/external/proxies/SignedOperationProxy.sol]: "msg.sender if no _signature_"?

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 15]
```diff
@@ -0,0 +1,550 @@
...
+
+            // require that the message is not expired
+            Require.that(
+                auth.expiration == 0 || auth.expiration >= block.timestamp,
+                FILE,
+                "Signed operation is expired",
+                p
+            );
+
+            // require that the sender is the sender
```
Comment (AntonioJuliano) [inline contracts/external/proxies/SignedOperationProxy.sol]: nit: "require the sender is correct"

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 16]
```diff
@@ -0,0 +1,550 @@
...
+                actions,
+                auth,
+                authStartIndex
+            );
+
+            // cache the index of the first action after this auth
+            uint256 authEndIndex = authStartIndex.add(auth.numActions);
+
+            // loop over all actions for which this auth applies
+            for (uint256 a = authStartIndex; a < authEndIndex; a++) {
```
Comment (AntonioJuliano) [inline contracts/external/proxies/SignedOperationProxy.sol]: if `authEndIndex` (`a`) is out of bounds for `actions` solidity will throw an error right?

File: src/types.ts [hunk 17]
```diff
@@ -59,7 +65,7 @@ export interface ContractConstantCallOptions extends Tx {
 }
 
 export interface AccountOperationOptions {
-  usePayableProxy?: boolean;
+  proxy?: ProxyType;
```
Comment (AntonioJuliano) [inline src/types.ts]: picking up the conversation on breaking changes there. It's almost weirder to have it not be a breaking change for javascript but be a breaking change for typescript. How about we just have both `proxy` and `usePayableProxy` (you can put a deprecated comment on it if you want)

File: src/modules/operate/AccountOperation.ts (hunk 3 above)
Comment (AntonioJuliano) [reply src/modules/operate/AccountOperation.ts]: maybe put some comments on this function as it's not super obvious especially as we're already in a class called `AccountOperation` and it's not obvious what "creating" another "operation" would do

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 18]
```diff
@@ -0,0 +1,550 @@
...
+        // create null account
+        Account.Info memory nullAccount = Account.Info({ owner: address(0), number: 0 });
+
+        // for each action that corresponds to the auth
+        for (uint256 i = 0; i < auth.numActions; i++) {
+            Actions.ActionArgs memory action = actions[startIndex + i];
+
+            // if action type has no second account, assume null account
+            Account.Info memory otherAccount =
+                (Actions.getAccountLayout(action.actionType) == Actions.AccountLayout.OnePrimary)
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: Should also be not true for any action with 2 accounts:
- Trade
- Liquidate
- Vaporize

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 19]
```diff
@@ -0,0 +1,550 @@
...
+            action.actionType,
+            primaryAccount.owner,
+            primaryAccount.number,
+            getAssetAmountHash(action.amount),
+            action.primaryMarketId,
+            action.secondaryMarketId,
+            action.otherAddress,
+            secondaryAccount.owner,
+            secondaryAccount.number,
+            keccak256(action.data)
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: Hashes the entire array. Right now for example the pattern `keccak256(abi.encode(...))` is very common

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 20]
```diff
@@ -0,0 +1,550 @@
...
+            uint256 authEndIndex = authStartIndex.add(auth.numActions);
+
+            // loop over all actions for which this auth applies
+            for (uint256 a = authStartIndex; a < authEndIndex; a++) {
+                // validate primary account
+                Actions.ActionArgs memory action = actions[a];
+                validateAccountOwner(accounts[action.accountId].owner, signer);
+
+                // validate second account in the case of a transfer action
+                if (action.actionType == Actions.ActionType.Transfer) {
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: There are three action types so it's not exactly the same, but yeah I'll do the equivalent fix

[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: I really only want it to be for `Transfer` which is the only action where the sender must be able to control both accounts. In the case of `Trade`, the `autoTrader` must be the operator instead, and in the case of `Liquidate` and `Vaporize` they do not have to be the operator of the second account.

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 21]
```diff
@@ -0,0 +1,550 @@
...
+        Actions.ActionArgs[] memory actions,
+        Authorization memory auth,
+        uint256 startIndex
+    )
+        private
+        pure
+        returns (bytes32)
+    {
+        // store in-memory sequential hash of each action
+        bytes memory actionsBytes = new bytes(auth.numActions.mul(32));
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: This bytes array must have 32bytes for each action, so it's longer than 32 bytes usually

File: contracts/external/proxies/SignedOperationProxy.sol [hunk 22]
```diff
@@ -0,0 +1,550 @@
...
+            bytes32 actionHash = getActionHash(
+                action,
+                accounts[action.accountId],
+                otherAccount
+            );
+
+            // store actionHash in the actionBytes array
+            /* solium-disable-next-line security/no-inline-assembly */
+            assembly {
+                mstore(add(actionsBytes, mul(0x20, add(i, 1))), actionHash)
```
[the remark this answers is not in the document]
Comment (BrendanChou) [reply contracts/external/proxies/SignedOperationProxy.sol]: I was able to make it non assembly 😄

**Pull request merged**